src/app/clusters/access-control-server/access-control-server.cpp

```diff
@@ -393,7 +393,10 @@ CHIP_ERROR LogAccessControlEvent(const AccessControl::Entry & entry, const Acces
     {
         for (size_t i = 0; i < subjectCount; ++i)
         {
-            ReturnErrorOnFailure(entry.GetSubject(i, subjectBuffer[i]));
+            Subject subject;
+            ReturnErrorOnFailure(entry.GetSubject(i, subject.nodeId));
```

Comment:
Would be better to just have a separate NodeId that we do the get into, instead of reusing `subject`.

That said, I don't understand why we are doing this at all, and the AccessControl.h documentation for GetSubject (which does not exist) does not explain it either.....

Reply:
This is purely for transcoding what you pass into the cluster into what the system module uses. As such, it won't be documented in the system module header, because it doesn't care about these details. Practically speaking, this is encoding a group ID in a node ID. There's a utility subject struct here to hold the necessary auth mode and overload a Convert function so it can have it when it does the conversion.

Reply:
My point is this:

1. It's not clear what the thing that GetSubject() returns contains.  That should be documented.
2. The input here is an AccessControl::Entry.  When we do GetSubject() on this, what does it actually produce?  It claims to produce a NodeId, so presumably for a group subject it would produce a group node id?
3. We then stick this in the .nodeId of a `Subject`, which seems ok, fine, we have a NodeId and we're putting it in there.
4. We do the Convert, then we put the `.nodeId` of the `Subject`, which now is holding something that is not at all a `NodeId` into `subjectBuffer` which claims to hold `NodeId` instances.
5. Then we treat this buffer of NodeId as a List of `uint64_t` when we do `staging.subjects.SetNonNull(subjectBuffer, subjectCount);`.  This happens to work because NodeId is a typedef, not a class, which is a bug we should really fix....

Here's what I think we should be doing instead so someone has a hope of following this code without doing a lot of digging (which is what I had to do just now to figure out why this is right after all):

* subjectBuffer should be an array of `uint64_t` (or SubjectId, I guess, if we want to add a typedef for that).
* `Subject` should have a `uint64_t` (or SubjectId), not a `NodeId` member, and it should be called `subjectId`, not `nodeId`.
* The `NodeId` we use to `GetSubject()` into should be a separate variable, not the member inside `Subject`.

Reply:
Sure, can I do all this in a followup?

Reply:
Can you please have the separate NodeId that we GetSubject() into in this PR and the rest as a followup?

Reply:
OK I can do that.

Reply:
NodeId broken out in latest revision, issue #15809 created for the rest.